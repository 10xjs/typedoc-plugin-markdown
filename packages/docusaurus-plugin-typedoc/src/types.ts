export interface PluginOptions {
  inputFiles: string[];
  docsRoot: string;
  out: string;
  sidebar: SidebarOptions | null;
  plugin: string[];
  disableOutputCheck?: boolean;
}

export interface SidebarOptions {
  fullNames: boolean;
  sidebarFile: string;
  globalsLabel: string;
  readmeLabel: string;
}

export interface Sidebar {
  [sidebarId: string]: SidebarItem[];
}

export interface SidebarCategory {
  type: string;
  label: string;
  items: SidebarItem[];
}

export type SidebarItem = SidebarCategory | string;

export interface LoadedContent {
  navigation: any;
}
