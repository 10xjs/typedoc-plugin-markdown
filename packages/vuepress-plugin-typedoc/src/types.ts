import { Application, ProjectReflection } from 'typedoc';

export interface PluginOptions {
  id?: string;
  logger?: string;
  inputFiles?: string[];
  out?: string;
  hideBreadcrumbs: boolean;
  hideProjectName: boolean;
  sidebar?: SidebarOptions | null;
}

export interface SidebarOptions {
  fullNames: boolean;
  parentCategory: string;
}

export interface LoadedContent {
  app: Application;
  project: ProjectReflection;
}
