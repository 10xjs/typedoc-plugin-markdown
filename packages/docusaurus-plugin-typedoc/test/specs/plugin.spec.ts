import * as path from 'path';

import { LoadContext } from '@docusaurus/types';

const mockAppBootstrap = jest.fn();
const mockAppConvert = jest.fn().mockImplementation(() => true);
const mockExpandInputFiles = jest.fn();
const mockGenerateDocs = jest.fn();
const mockWriteSidebar = jest.fn();

jest.mock('typedoc', () => {
  return {
    Application: jest.fn().mockImplementation(() => {
      return {
        bootstrap: mockAppBootstrap,
        convert: mockAppConvert,
        generateDocs: mockGenerateDocs,
        expandInputFiles: mockExpandInputFiles,
        renderer: {
          addComponent: jest.fn(),
          getComponent: jest.fn().mockImplementation(() => {
            return {
              isOutputDirectory: jest.fn().mockImplementation(() => true),
              getNavigation: jest.fn().mockImplementation(() => []),
            };
          }),
        },
      };
    }),
  };
});

jest.mock('../../dist/components/front-matter-component');

jest.mock('../../dist/sidebar', () => {
  return {
    writeSidebar: mockWriteSidebar,
  };
});

describe(`Plugin:`, () => {
  let plugin;
  const context = { siteDir: '/siteDir' } as LoadContext;
  const defaultOptions = {
    plugin: ['typedoc-plugin-markdown'],
    theme: path.resolve(__dirname, '..', '..', 'dist', 'theme'),
  };
  beforeEach(() => {
    mockAppBootstrap.mockClear();
    mockGenerateDocs.mockClear();
    mockExpandInputFiles.mockClear();
    mockWriteSidebar.mockClear();
    jest.isolateModules(() => {
      plugin = require('../../dist/plugin').default;
    });
  });

  describe(`(typedoc options):`, () => {
    test('should bootstrap with default options', async () => {
      await plugin(context, {});
      expect(mockAppBootstrap).toHaveBeenCalledWith(defaultOptions);
    });

    test('should bootstrap with additional options', async () => {
      await plugin(context, { mode: 'file' });
      expect(mockAppBootstrap).toHaveBeenCalledWith({
        ...defaultOptions,
        mode: 'file',
      });
    });

    test('should bootstrap with additional typedoc plugin', async () => {
      await plugin(context, { plugin: ['typedoc-plugin-xyz'] });
      expect(mockAppBootstrap).toHaveBeenCalledWith({
        ...defaultOptions,
        plugin: ['typedoc-plugin-markdown', 'typedoc-plugin-xyz'],
      });
    });

    test('should not duplicate markdown plugin', async () => {
      await plugin(context, {
        plugin: ['typedoc-plugin-markdown', 'typedoc-plugin-xyz'],
      });
      expect(mockAppBootstrap).toHaveBeenCalledWith({
        ...defaultOptions,
        plugin: ['typedoc-plugin-markdown', 'typedoc-plugin-xyz'],
      });
    });
  });
  describe(`(plugin options):`, () => {
    test('should set default input files', async () => {
      await plugin(context, {});
      expect(mockExpandInputFiles).toHaveBeenCalledWith(['../src/']);
    });
    test('should set custom input files', async () => {
      await plugin(context, { inputFiles: ['../../lib'] });
      expect(mockExpandInputFiles).toHaveBeenCalledWith(['../../lib']);
    });
    test('should set default output directory', async () => {
      await plugin(context, {});
      expect(mockGenerateDocs).toHaveBeenCalledWith(
        true,
        context.siteDir + '/docs/api',
      );
    });
    test('should set custom output directory', async () => {
      await plugin(context, { out: 'myapi' });
      expect(mockGenerateDocs).toHaveBeenCalledWith(
        true,
        context.siteDir + '/docs/myapi',
      );
    });
    test('should set in root dir', async () => {
      await plugin(context, { out: '' });
      const outputDirectory = context.siteDir + '/docs';
      expect(mockGenerateDocs).toHaveBeenLastCalledWith(true, outputDirectory);
    });
  });

  describe(`(sidebar options):`, () => {
    test('should write sidebar with default options', async () => {
      await plugin(context, {});
      expect(mockWriteSidebar).toHaveBeenLastCalledWith(
        true,
        context.siteDir,
        'api',
        {
          fullNames: false,
          sidebarFile: 'typedoc-sidebar.js',
          globalsLabel: "Globals",
          readmeLabel: "README"
        },
        [],
      );
    });
    test('should write sidebar with merge sidebar options', async () => {
      await plugin(context, {
        sidebar: { sidebarFile: 'custom-sidebar.js' },
      });
      expect(mockWriteSidebar).toHaveBeenLastCalledWith(
        true,
        context.siteDir,
        'api',
        {
          fullNames: false,
          sidebarFile: 'custom-sidebar.js',
          globalsLabel: "Globals",
          readmeLabel: "README"
        },
        [],
      );
    });

    test('should not write sidebar if null', async () => {
      await plugin(context, { sidebar: null });
      expect(mockWriteSidebar).toHaveBeenCalledTimes(0);
    });
  });
});
