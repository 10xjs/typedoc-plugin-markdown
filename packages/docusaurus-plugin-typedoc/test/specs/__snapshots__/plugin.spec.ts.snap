// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin: (options) should merge another plugin 1`] = `
Array [
  "typedoc-test-plugin-1",
  "typedoc-test-plugin-2",
  "typedoc-plugin-markdown",
]
`;

exports[`Plugin: (options) should not duplicate markdown plugin 1`] = `
Array [
  "typedoc-test-plugin",
  "typedoc-plugin-markdown",
]
`;

exports[`Plugin: (options) should set default plugin to 'typedoc-plugin-markdown' 1`] = `
Array [
  "typedoc-plugin-markdown",
]
`;

exports[`Plugin: (render) should write docs 1`] = `
Array [
  "globals.md",
  "index.md",
  "interfaces",
  "modules",
]
`;

exports[`Plugin: (sidebars) should update sidebars.js' 1`] = `
"module.exports = {
  \\"someSidebar\\": {
    \\"Docusaurus\\": [
      \\"doc1\\"
    ],
    \\"Features\\": [
      \\"mdx\\"
    ]
  },
  \\"typedoc\\": [
    {
      \\"type\\": \\"category\\",
      \\"label\\": \\"none\\",
      \\"items\\": [
        \\"api/index\\",
        \\"api/globals\\",
        {
          \\"type\\": \\"category\\",
          \\"label\\": \\"Modules\\",
          \\"items\\": [
            \\"api/modules/_theme_\\"
          ]
        },
        {
          \\"type\\": \\"category\\",
          \\"label\\": \\"Interfaces\\",
          \\"items\\": [
            \\"api/interfaces/_theme_.interfaceitema\\"
          ]
        }
      ]
    }
  ]
};"
`;
